### ISSUE TITLE
[CWM] Mykomaps with Leaflet deep dive

### ISSUE THREAD

**Issue by lin-d-hop** [DigitalCommons/mykomap#259]
** Track under the Cooperative World Map project **

## Description

This issue is a place to document the findings of a deep dive to unpack how much work is required to get a functional MykoMap for the CWM project with Leaflet.

By 'functional MykoMap' the criteria we are using is:
- 500k markers
- Loads map and markers in 2 seconds
- Using your own desktop (assuming it is not under 50 tabs duress)
- Ask yourself 'If I can a mostly disinterested user on this site, is this slow enough that I'd just close it out of frustration?'

The task is a **2 day timeboxed** exploration of the work it will take to create a functional MM using Leaflet. Building on the analysis done in #256 #257 and #254 the goal is try and understand as deeply as possible the work involved in streamlining MM as much as possible, such that we can use Leaflet to deliver a functional MM.

The idea is that this get's a bit hacky. Make forks, cull functionality, mock backend calls. Play in the sandbox.

## Acceptance Criteria
- As much understanding as possible about what the slowest and messiest parts of MM are
- As much understanding as possible about where the biggest wins for refactoring will be found wrt delivering a functional map
- As detailed a breakdown of tasks and plan of attack as possible within the timebox
- Rough estimates on complexity. Tshirt sizes will be okay, but go as granular as time allows.

A word of caution - it might be tempting to think only breadth first (what are all the options) or only depth first (how do I get this one option perfect). The real skill of this task is to balance between both perspectives. You might think of them as two different voices in your brain. All them both to the table in equal measure :)

**Comment by wu-lee** [comment]
Some notes added here: https://github.com/DigitalCommons/leaflet-markercluster-vs-supercluster/blob/main/NOTES.md

TL;DR:

The main bottlenecks:
- loading data (somewhat, depending on network)
- parsing and processing data - this can be slow for large datasets
  - unintuitively, binary formats are not necessarily a win, because of native browser JSON optimisations
  - except perhaps something like Google's Protocol Buffers?
- cluster indexing
  - esp. slow with Leaflet's MarkerCluster
  - Supercluster is much better
    - Works with 500k pins quickly once indexed
    - Significant index delay, but perhaps acceptable on modern browsers (very small number of seconds)
    - but: requires re-indexing whenever visible pins change
    - and: does not come with a lot of MarkerCluster's functionality, (spidering, reclustering on pan/zoom)
    - this missing functionality can be re implemented without too much effort
- searching and filtering
  - text search especially
  - category filtering

Workarounds:
- Use JSON, but judiciously minimised and abbreviated for initial load
  - Prefer arrays to objects
  - avoid nesting
  - avoid repetition
  - use short keys on objects, 
  - use integers instead of URIs and textual identifiers for categories
  - use bitfields if possible for multi-value categories
  - avoid unnecessarily long decimals in lat/lng (4dp should be ample)
- Avoid to much post-processing of data once loaded
- Avoid triggering GC (through unnecessary data churn)
- Load supplemental data on demand (extended pin information like descriptions)
- Enable HTTP/2 on the server, with transport compression
- Google's [Protobuffers][2] might offer improvements but require more developer effort.to employ

Overall: start with minimal functionality and add new functions carefully with ["benchmark driven development"][1] combined with profiling. Essentially this would be a new product, as much of Mykomaps would need reworking.

Avoiding the client-side cluster-indexing cost may be hard to avoid, as it needs to be done whenever the visible pins change, the index is very large, and a wide range of possible filter permutations make pre-computation harder. But reclustering once indexed is relatively cheap.

Server-side clustering could be done but then the server can become the bottleneck: we would need new server infrastructure. Even then, every zoom and pan then becomes more costly as visible pins would need to be re-downloaded from the server on each zoom/pan. Prototyping would be needed to get idea of whether this cost is acceptable. Something like server-sent-events might reduce query round-trips, and caching might reduce the amount of data, but caching is notoriously tricky and and a source of bugs.

[1]: https://github.com/felixge/faster-than-c
[2]: https://auth0.com/blog/beating-json-performance-with-protobuf/
